torch/optim/_multi_tensor/sgd.py

```diff
@@ -105,51 +101,21 @@ def step(self, closure=None):
                     if p.grad.is_sparse:
                         has_sparse_grad = True
 
-                        if momentum != 0:
+                        if group['momentum'] != 0:
                             raise RuntimeError('SGD does not support momentum for sparse gradients')
 
             if grads == []:
                 return loss
 
-            if weight_decay != 0:
-                grads = torch._foreach_add(grads, params_with_grad, alpha=weight_decay)
-
-            if momentum != 0:
-                bufs = []
-
-                all_states_with_momentum_buffer = True
-                for i in range(len(states)):
-                    if 'momentum_buffer' not in states[i]:
-                        all_states_with_momentum_buffer = False
-                        break
-                    else:
-                        bufs.append(states[i]['momentum_buffer'])
-
-                if all_states_with_momentum_buffer:
-                    torch._foreach_mul_(bufs, momentum)
-                    torch._foreach_add_(bufs, grads, alpha=1 - dampening)
-                else:
-                    bufs = []
-                    for i in range(len(states)):
-                        if 'momentum_buffer' not in states[i]:
-                            buf = states[i]['momentum_buffer'] = torch.clone(grads[i]).detach()
-                        else:
-                            buf = states[i]['momentum_buffer']
-                            buf.mul_(momentum).add_(grads[i], alpha=1 - dampening)
-
-                        bufs.append(buf)
-
-                if nesterov:
-                    torch._foreach_add_(grads, bufs, alpha=momentum)
-                else:
-                    grads = bufs
-
-            if not has_sparse_grad:
-                torch._foreach_add_(params_with_grad, grads, alpha=-group['lr'])
-            else:
-                # foreach APIs dont support sparse
-                for i in range(len(params_with_grad)):
-                    params_with_grad[i].add_(grads[i], alpha=-group['lr'])
+            F.sgd(params_with_grad,
```

Comment:
What remains of the `optim._multi_tensor.SGD()` class seems to be exactly the same as the `optim.SGD()` one.
Do you think we could refactor this code to have a single implementation (and maybe a new attribute that hold the functional step function?

Reply:
This is actually very good point, and i would love to have further discussions about this. But currently there are certain differences with multi_tensor and non-multi_tensor optimizers. For example: 

For SGD we have :  https://github.com/pytorch/pytorch/blob/bfe03120eebc8b2b7c256d08887a782de9b6177c/torch/optim/sgd.py#L120

For ASGD: https://github.com/pytorch/pytorch/blob/bfe03120eebc8b2b7c256d08887a782de9b6177c/torch/optim/asgd.py#L88

Which we don't need them in multi-tensor case. We will return back to functional optimizers when discussions of functional Optimizers finalizes in Vmap side. Maybe we can investigate your suggestion in that time ?

Reply:
Ok